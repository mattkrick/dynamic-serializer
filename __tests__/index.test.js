import DynamicSerializer from '../src/index';

test('handles top-level paths', () => {
  const paths = ['top'];
  const snapshot = {
    top: 'rand'
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles deep-nested paths', () => {
  const paths = ['top.two.three'];
  const snapshot = {
    top: {
      two: {
        three: 'rand'
      }
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles nested arrays', () => {
  const paths = ['top.two.three'];
  const snapshot = {
    top: {
      two: {
        three: ['rand1', 'rand2']
      }
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles single value in nested arrays', () => {
  const paths = ['top.two.three.1'];
  const snapshot = {
    top: {
      two: {
        three: ['static', 'rand2']
      }
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('reuses pre-established vars for other snapshots', () => {
  const paths = ['top.two.three.1'];
  const snapshot = {
    top: {
      two: {
        three: ['static', 'rand2']
      }
    }
  };

  const paths2 = ['foreignKey'];
  const snapshot2 = {
    foreignKey: snapshot.top.two.three[1]
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  dynamicSerializer.toStatic(snapshot2, paths2);
  expect(snapshot).toMatchSnapshot();
  expect(snapshot2).toMatchSnapshot();
  expect(snapshot.top.two.three[1]).toBe(snapshot2.foreignKey);
});

test('does not replace null or undefined', () => {
  const paths = ['top.two.three.1', 'top.two.three.0'];
  const snapshot = {
    top: {
      two: {
        three: [null, undefined]
      }
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('does not replace null or undefined or missing', () => {
  const paths = ['top.two.three.1', 'top.two.three.0', 'top.two.4', 'top.two.5'];
  const snapshot = {
    top: {
      two: {
        three: [null, undefined]
      },
      4: undefined
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('goes deep after a single array point', () => {
  const paths = ['top.a', 'top.two.three.1.a'];
  const snapshot = {
    top: {
      a: 'rand',
      two: {
        three: [
          null,
          {
            a: 'rand'
          }
        ]
      }
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles an array in an array', () => {
  const paths = ['top.two.three.1.2'];
  const snapshot = {
    top: {
      two: {
        three: [null, ['static', 'static', 'rand']]
      }
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles a top level array', () => {
  const paths = ['1.0'];
  const snapshot = ['static', [
    'rand'
  ]];
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles final-path empty values', () => {
  const paths = ['a.b', 'a.c'];
  const snapshot = {
    a: {
      b: '',
      c: [null]
    }
  };
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});

test('handles an array of objects', () => {
  const paths = ['a.b', 'a.c'];
  const snapshot = [
    {
      a: {
        b: 'rand',
      }
    }
  ];
  const dynamicSerializer = new DynamicSerializer();
  dynamicSerializer.toStatic(snapshot, paths);
  expect(snapshot).toMatchSnapshot();
});
