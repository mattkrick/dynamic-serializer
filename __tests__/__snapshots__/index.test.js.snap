// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not replace null or undefined 1`] = `
Object {
  "top": Object {
    "two": Object {
      "three": Array [
        null,
        undefined,
      ],
    },
  },
}
`;

exports[`does not replace null or undefined or missing 1`] = `
Object {
  "top": Object {
    "4": undefined,
    "two": Object {
      "three": Array [
        null,
        undefined,
      ],
    },
  },
}
`;

exports[`goes deep after a single array point 1`] = `
Object {
  "top": Object {
    "a": 0,
    "two": Object {
      "three": Array [
        null,
        Object {
          "a": 0,
        },
      ],
    },
  },
}
`;

exports[`handles a top level array 1`] = `
Array [
  "static",
  Array [
    0,
  ],
]
`;

exports[`handles an array in an array 1`] = `
Object {
  "top": Object {
    "two": Object {
      "three": Array [
        null,
        Array [
          "static",
          "static",
          0,
        ],
      ],
    },
  },
}
`;

exports[`handles an array of objects 1`] = `
Array [
  Object {
    "a": Object {
      "b": 0,
    },
  },
]
`;

exports[`handles deep-nested paths 1`] = `
Object {
  "top": Object {
    "two": Object {
      "three": 0,
    },
  },
}
`;

exports[`handles final-path empty values 1`] = `
Object {
  "a": Object {
    "b": "",
    "c": Array [
      null,
    ],
  },
}
`;

exports[`handles nested arrays 1`] = `
Object {
  "top": Object {
    "two": Object {
      "three": Array [
        0,
        1,
      ],
    },
  },
}
`;

exports[`handles single value in nested arrays 1`] = `
Object {
  "top": Object {
    "two": Object {
      "three": Array [
        "static",
        0,
      ],
    },
  },
}
`;

exports[`handles top-level paths 1`] = `
Object {
  "top": 0,
}
`;

exports[`reuses pre-established vars for other snapshots 1`] = `
Object {
  "top": Object {
    "two": Object {
      "three": Array [
        "static",
        0,
      ],
    },
  },
}
`;

exports[`reuses pre-established vars for other snapshots 2`] = `
Object {
  "foreignKey": 0,
}
`;
